File: packages/shared/src/session.ts
import { z } from 'zod';

export const ChoiceSchema = z.object({
    id: z.string(),
    text: z.string(),
    next_scene_id: z.string(),
    correctness: z.boolean().optional(),
    feedback: z.string(),
});

export const HotspotSchema = z.object({
    id: z.string(),
    label: z.string(),
    x: z.number(),
    y: z.number(),
    w: z.number(),
    h: z.number(),
    dialogue: z.string(),
    expand_text: z.string(),
    sound_cue: z.string().optional(),
});

export const QuizSchema = z.object({
    question: z.string(),
    options: z.array(z.string()),
    correct_index: z.number(),
    explanation: z.string(),
});

export const SceneSchema = z.object({
    id: z.string(),
    narration: z.string(),
    image_prompt: z.string().optional(),
    choices: z.array(ChoiceSchema),
    hotspots: z.array(HotspotSchema),
    quiz: QuizSchema.optional(),
});

export const SessionStateSchema = z.object({
    current_scene_index: z.number(),
    score: z.number(),
    decisions: z.array(z.string()),
});

export const SessionSchema = z.object({
    id: z.string(),
    created_at: z.string().datetime(),
    topic: z.string(),
    settings: z.record(z.any()).optional(),
    scenes: z.array(SceneSchema),
    state: SessionStateSchema,
});

export type Choice = z.infer<typeof ChoiceSchema>;
export type Hotspot = z.infer<typeof HotspotSchema>;
export type Quiz = z.infer<typeof QuizSchema>;
export type Scene = z.infer<typeof SceneSchema>;
export type SessionState = z.infer<typeof SessionStateSchema>;
export type Session = z.infer<typeof SessionSchema>;

File: packages/shared/python/session.py
from typing import List, Optional, Any, Dict
from pydantic import BaseModel, Field
from datetime import datetime

class Choice(BaseModel):
    id: str
    text: str
    next_scene_id: str
    correctness: Optional[bool] = None
    feedback: str

class Hotspot(BaseModel):
    id: str
    label: str
    x: float
    y: float
    w: float
    h: float
    dialogue: str
    expand_text: str
    sound_cue: Optional[str] = None

class Quiz(BaseModel):
    question: str
    options: List[str]
    correct_index: int
    explanation: str

class Scene(BaseModel):
    id: str
    narration: str
    image_prompt: Optional[str] = None
    choices: List[Choice]
    hotspots: List[Hotspot]
    quiz: Optional[Quiz] = None

class SessionState(BaseModel):
    current_scene_index: int
    score: int
    decisions: List[str]

class Session(BaseModel):
    id: str
    created_at: datetime
    topic: str
    settings: Optional[Dict[str, Any]] = Field(default_factory=dict)
    scenes: List[Scene]
    state: SessionState
